bootstore/src/schemes/v0/state_initial_member.rs

```diff
@@ -0,0 +1,411 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! FSM API for `State::InitialMember`
+
+use std::collections::BTreeMap;
+
+use crate::schemes::v0::fsm_output::{ApiError, ApiOutput};
+use crate::trust_quorum::{LearnedSharePkgV0, RackSecret, SharePkgV0};
+
+use super::fsm::StateHandler;
+use super::fsm_output::Output;
+use super::messages::{Envelope, Error, RequestType, Response, ResponseType};
+use super::state::{
+    FsmCommonData, RackInitState, RequestMetadata, ShareIdx, State,
+};
+use secrecy::{ExposeSecret, Secret};
+use sled_hardware::Baseboard;
+use uuid::Uuid;
+
+#[derive(Debug)]
+pub struct InitialMemberState {
+    pub pkg: SharePkgV0,
+
+    /// Shares given to other sleds. We mark them as used so that we don't
+    /// hand them out twice. If the same sled asks us for a share, because
+    /// it crashes or there is a network blip, we will return the same
+    /// share each time.
+    ///
+    /// Note that this is a fairly optimistic strategy as the requesting
+    /// sled can always go ask another sled after a network blip. However,
+    /// this guarantees that a single sled never hands out more than one of
+    /// its shares to any given sled.
+    ///
+    /// We can't do much better than this without some sort of centralized
+    /// distributor which is part of the reconfiguration mechanism in later
+    /// versions of the trust quourum protocol.
+    pub distributed_shares: BTreeMap<Baseboard, ShareIdx>,
+
+    // Acknowledgements tracked during rack initialization if this node is
+    // the one acting as coordinator (local to RSS).
+    pub rack_init_state: Option<RackInitState>,
```

Comment:
Does this remain relevant for the entire life-cycle of the `InitialMemberState`?  Otherwise, I'm wondering if you want a special `RackCoordinatorState`, which then becomes `InitialMemberState` (without this field) once RSS is done.

Reply:
I understand the suggestion for `RackCoordinatorState` and I considered it. But I don't think it makes sense as part of the protocol as a whole. The reason is that once a rack init starts it *must* succeed. If it fails, we wipe all sleds and start again via an RSS reset. There is no two-phase commit type thingy, and so as soon as the api request for `init_rack` comes in it moves directly to `InitialMember` state. This is symmetric with what peers do when they receive a message containing `RequestType::Init(pkg)`. If we had an extra state at the coordinator, peers would then move directly to `InitialMember` state and would be able to answer requests while the coordinator was lagging behind them. One would think that we could then add an extra state to the peers such as `PreparedInitialMember`, but that actually reduces reliability. You could envision a scenario where the following happens and the rack is completely unavailable:

 1. Coordinator completes RSS after hearing `InitAck` from all peers and responds to its API with success. 
 2. RSS thinks rack is initialized
 3. Coordinator crashes before sending `CommitRackInit` messages to peers.
 4. Peers are struck in `PreparedInitialMember` state and the rack is unavailable

Now, you may be tempted to fix the above scenario by having the coordinator send a `CommitRackInit` to all peers and wait for acks before responding success to the API. But, then we are back at peer nodes answering requests before the coordinator again. And it's also possible that not all nodes ack the second time, since one of them is down temporarily. So there isn't really a benefit here.